### PR TITLE
Enabling ATen Distribution kernels for AARCH64 using OpenRNG

### PR DESCRIPTION
**Pull request by agrawal-aka**
_**Enables Distribution ATen kernel for ARM CPU's using OpenRNG Library.**_
**_Enhances performance of bernoulli and exponential distribution operations for ARM CPU's._**
(https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/DistributionKernels.cpp)

Contribution:

- With this PR, we propose to add OpenRNG to the pytorch/third_party folder as a submodule and support static linkage of OpenRNG : (libopenrng.a)

- USE_OPENRNG is implemented as cmake_dependent_option, which is available only on AARCH64, and by default is set to ON.

- We also give user the flexibility to USE_SYSTEM_OPENRNG (similar to USE_SYSTEM_SLEEF), to dynamically link libopenrng.so to libtorch_cpu.so, if user sets the OPENRNG_ROOT_DIR environment variable before build.

The enablement summary is captured below:
Screenshot 2024-11-26 135226

This change targets performance enhancement for distribution kernels (Bernoulli and Exponential) for ARM. The primary use case targeted is DL/LLM training, where Bernoulli kernel is used in dropout implementation, to randomly drop neurons during weight updation to prevent overfitting.

This change leverages OpenRNG, which uses neon backend for RNG (random number generator), and utilises same function calls as MKL VSL, which provides the observed performance boost over the native aten implementation route currently followed for ARM.

**OP level performance boost:**
Example: Sampled random input tensors of shapes from popular LLM models.
image

**Model level performance boost:**
Example: BERT training for single epoch.
Screenshot 2024-09-03 111921

Note: GR- Graviton 3 Perf Instance

**Insights:**

1. For Bernoulli kernel using OpenRNG, performance now scales up with increasing cores for GR instance. (not happening previously, and over ~100x performance gain for largest shapes is observed. (Table 1)
2. For Bernoulli op kernel using OpenRNG, significant drop in avg CPU time (~32x improvement) during DL Training (BERT) is achieved. (Table 2)
3. Similar observations are also made for exponential kernel.

cc @msaroufim @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @milpuz01

### CODE REVIEW

**Comment by maajidkhann** [comment]
@snadampal @milpuz01 Can you please review this PR whenever you get some time.

**Review by albanD — CHANGES_REQUESTED** [review]
I expect most of the speedup is from multithreading the sampling which could also be done with the regular kernel.
We don't do this to provide some stability of the sampling but I don't recall the full details.
Also we had other issues with MKL sampling where the randomness was not enough. So this will need to be reviewed very carefully.

cc @malfet do you know who's the best person to handle this?

**Comment by agrawal-aka** [comment]
Hello @albanD, with the current change, significant speedup is achieved even for a single core as well. As highlighted, with the change we are able to leverage neon implementation of RNG engines. The performance numbers for single core are also attached for tensor shapes from BERT model for bernoulli op, showing upto ~4x speedup for largest shapes. 

image

With parallelization using multiple cores, we achieve further performance boost as already mentioned in the main comment.

Also, for reference, a recent blog published on ARM website talks about leveraging openrng (part of ARMPL) for accelerating bernoulli operator in PyTorch: 
https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/openrng-24-04

**Comment by maajidkhann** [comment]
@robert-hardwick Hello Robert, can you please add the below CI label to this PR.
**"ciflow/linux-aarch64"**

**Comment by maajidkhann** [comment]
CC @malfet . Can you please review this PR and provide your feedback whenever you get some time.

**Comment by maajidkhann** [comment]
The armpl binary future releases will continue to have VSL RNGs from OpenRNG making this integration more stable.

image

**Source:** https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/openrng-24-04

**Comment by agrawal-aka** [comment]
**Previous Description of this PR :** 
Armpl_comment_1
armpl_comment_2

**Comment by agrawal-aka** [comment]
The PR is currently updated to integrate OpenRNG instead of ARMPL, for performance enhancement of distribution kernels for AARCH64, which avoids having two BLAS libraries in PyTorch as suggested as feedback.

**Comment by malfet** [comment]
@agrawal-aka in addition to performance numbers, would it be possible to add some sort of benchmark about distribution properties of the proposed generators?
And is there a mirror of the repository on GitHub? If not, can it be integrated into docker images, as Gitlab is often down

**Comment by chrisgoodyer** [comment]
> And is there a mirror of the repository on GitHub? If not, can it be integrated into docker images, as Gitlab is often down

Hi.  No, Arm do not have a mirror of this repository on GitHub.

When you say "Gitlab is often down" do you mean the regular www.gitlab.com site, or the Arm specific one (gitlab.arm.com) that is run by Arm?  If it's the latter then any evidence you've seen of that being the case would be useful to gather to be able to talk to the relevant teams to get the quality of the service improved.

Thanks!

**Comment by agrawal-aka** [comment]
Hi @malfet, it is important to point out with regards to distribution quality of RNG:
1. The issue https://github.com/pytorch/pytorch/issues/132395 is still open. And currently buggy code exists in OSS for the MKL route. As this PR enables OpenRNG route utilising same APIs as MKL, the above integration suffers from the same quality issue until a fix is merged.

2. I applied the changes mentioned in this PR #132603 that fixes the issue #132395, and verified on both IL and graviton (with OpenRNG integration of this PR), and it seems to give 0 duplicates:

```
import torch

high_bits_for_seed = 16000000000000000000           # to use "good quality" seed
_ = torch.manual_seed (high_bits_for_seed + 2024)

prob = torch.ones (26)
dups_mult = 0
perm_counts_mult = {}
for _ in range (1_000_000):
    p = tuple (torch.multinomial (prob, prob.numel(), replacement=False).tolist())
    if  p in perm_counts_mult:
        dups_mult += 1
        perm_counts_mult[p] += 1
    else:
        perm_counts_mult[p] = 1

print ('duplicate multinomial perms: ', dups_mult)
print ('multiple multinomial perms:  ', (torch.tensor (list (perm_counts_mult.values())) > 1).sum().item())
print ('max of perm_counts_mult:     ', torch.tensor (list (perm_counts_mult.values())).max().item())
print ('len (perm_counts_mult):      ', len (perm_counts_mult))
```
```
duplicate multinomial perms:  0
multiple multinomial perms:   0
max of perm_counts_mult:      1
len (perm_counts_mult):       1000000
```

Essentially it does fix the randomness issue to a certain extent and atleast better than the current version in OSS, though not completely addressing the concern that seeding a local MT with a global MT would not be a best practice.

3. If either this fix https://github.com/pytorch/pytorch/pull/132603 or any other fix for the above issue (current discussion is in this thread: https://github.com/pytorch/pytorch/pull/132742) goes through, as OpenRNG utilises same APIs as MKL VSL, the accelerated route for ARM via OpenRNG would also be enabled via this PR https://github.com/pytorch/pytorch/pull/134942.

With regards to the concern:
image
can refer the comment by @chrisgoodyer

**Comment by fadara01** [comment]
Hi @agrawal-aka, thank you for your PR.
I tried to rebase and run the AArch64 Linux CI workflow but there are some merge conflicts.
Could you please resolve the conflicts s.t. we can run the full AArch64 CI against this PR?

**Comment by agrawal-aka** [comment]
Hi @fadara01, thanks for your comment, yes, I am working on it and will update in a while.

**Comment by agrawal-aka** [comment]
Hi @fadara01, I have resolved conflicts and rebased the PR, could you trigger the AArch64 Linux CI workflow?

**Comment by agrawal-aka** [comment]
Hi @fadara01, there was a lint fix needed which I have done, all other checks were passing, could you retrigger the CI? Thanks.
bernoulli_ci

**Comment by agrawal-aka** [comment]
Hi @fadara01, could you suggest how do we proceed ahead from here?

> Are you ok with the current flow for ARM going to openrng or if required we can also make it go to default path and merge this PR and once the issue of randomness is resolved in future, we can make it again go to openrng path and in Intel case, MKL path?

**Comment by CaoE** [comment]
> @nikhil-arm @agrawal-aka @maajidkhann @malfet I have a possible way to move forward with openrng in pytorch where we put the `vslStream` inside a new `MKLGeneratorImpl` (see https://github.com/michalowski-arm/pytorch/tree/mklrng). This way we can initialize the stream with a seed at the first call to either `exponential_` or `bernoulli_` and then keep using the same stream at subsequent calls, so we avoid the repeated sequences problem mentioned in #132395. It will also reenable MKL rng since the api is the same (@CaoE). I cannot open a PR on top this one though, since it's on a branch in a fork of pytorch. Would you mind pushing it to a new branch of pytorch directly? For now you can see proposed changes if you got to the mklrng branch of my fork above.

@michalowski-arm This is great! This would solve the repeated sequences problem.
Not sure if I understand it correctly. If the user resets the state of the CPU generator externally (for example, saving the state and then restoring it to produce the same sequence as a specific call), MK generator does not seem to be able to produce the specific sequence the user wants.

Do we need to set the state of MKL generator when setting the CPU generator ?

**Comment by michalowski-arm** [comment]
I pushed another commit which shows one possible way to link MKLGenerator state change to CPUGenerator state changes. It might need some ironing out but the approach should work in principle.

**Comment by michalowski-arm** [comment]
@agrawal-aka Are you fine with me opening a new PR which includes your two commits from this PR alongside my changes?

**Comment by agrawal-aka** [comment]
Hi @michalowski-arm, thanks for the update and resolution of the repeated sequences problem mentioned in #132395. I tried creating a branch directly onto Pytorch OSS, but access is denied. I have two suggestions in this regard:
 
1. As #132395 is present for both intel (MKL route, currently disabled in OSS) and arm (OpenRNG route, which is currently not merged in OSS, proposed in this PR), a separate bug fix PR on top of the OSS PyTorch (excluding any changes related to OpenRNG, because the randomness issue is independent of the same) can be raised, and that can be reviewed and merged separately, so that first the MKL route is enabled. 

Once it is established there is no randomness issue for x86 platform, this PR #134942 can be rebased over the latest main and subsequently merged, as the same API calls can be utilised using OpenRNG, would enable the distribution kernels for AARCH64, which was originally proposed in this PR.
 
2. The other option would be as this original PR #134942 is not merged yet, it would be good to have all the changes consolidated here including yours at one place. I can cherry pick your two commits and add it to this PR itself, and we can update the description and submit it as a unified OpenRNG integration and randomness bug fix PR. It makes it easy for reviewers as all previous context has been on this long-standing PR.

**Comment by michalowski-arm** [comment]
@agrawal-aka The first one might be best, I'll make some changes and open a PR which focuses only on reenabling the MKL path.

**Comment by agrawal-aka** [comment]
Sure, @michalowski-arm, that works well. I'll keep track of your PR and once it’s approved and merged, I’ll make the necessary minor changes for OpenRNG in the newly added files. After that, this PR #134942 can proceed separately. Thanks.

File: .gitmodules [hunk 1]
```diff
@@ -132,3 +132,6 @@
 [submodule "third_party/aiter"]
 	path = third_party/aiter
 	url = https://github.com/ROCm/aiter.git
+[submodule "third_party/openrng"]
+	path = third_party/openrng
+	url = https://git.gitlab.arm.com/libraries/openrng.git
```
Comment (malfet) [inline .gitmodules:137]: We can't use gitlab, as it is unstable

**Review by malfet — CHANGES_REQUESTED** [review]
We can not depen on anything hosted on gitlab, @milpuz01 should vividly remember all the infra problems we had with that repo when trying to integrate KliediAI)

**Comment by fadara01** [comment]
@agrawal-aka - https://github.com/pytorch/pytorch/pull/151218 is finally merged, so this PR is now unblocked.

We're also working internally on getting an OpenRNG github mirror, it should be available very soon

Are you still interested in working towards getting this PR in a mergable state?
If so, could you please rebase and address the merge conflicts?

If not, we'll cherry-pick your commits into a new PR and work towards getting that merged.

**Comment by agrawal-aka** [comment]
@fadara01 thanks for your relentless pursuit in getting https://github.com/pytorch/pytorch/pull/151218, merged, yes I am willing to get this PR in a mergable state, once we get the github mirror up, sure I'll rebase and address the merge conflicts

**Comment by agrawal-aka** [comment]
Hi @fadara01, seems https://github.com/pytorch/pytorch/pull/151218, got reverted again, could you recheck?

File: .gitmodules (hunk 1 above)
Comment (aditew01) [reply .gitmodules:137]: @agrawal-aka please point to  https://github.com/arm/openrng

**Comment by aditew01** [comment]
> Hi @fadara01, seems #151218, got reverted again, could you recheck?

the gaurd to fix the failing builds will be pushed shortly.

File: .gitmodules (hunk 1 above)
Comment (agrawal-aka) [reply .gitmodules:137]: sure, thanks @aditew01 for porting this repo to github, will update

**Comment by agrawal-aka** [comment]
> the gaurd to fix the failing builds will be pushed shortly.

Hi @aditew01 is there any update on #151218 ?
